File: Common/src/main/java/foundry/veil/anim/Frame.java
package foundry.veil.anim;

import foundry.veil.math.Easings;
import net.minecraft.world.phys.Vec3;

public class Frame {
    Vec3 position;
    Vec3 rotation;
    Vec3 scale;

    public Frame(Vec3 position, Vec3 rotation, Vec3 scale){
        this.position = position;
        this.rotation = rotation;
        this.scale = scale;
    }

    public Vec3 getPosition() {
        return position;
    }

    public Vec3 getRotation() {
        return rotation;
    }

    public Vec3 getScale() {
        return scale;
    }

    public Frame copy(){
        return new Frame(position, rotation, scale);
    }

    public Frame interpolate(Frame frame, float progress, Easings.Easing easing) {
        Vec3 position = this.position.add(frame.position.subtract(this.position).scale(easing.ease(progress)));
        Vec3 rotation = this.rotation.add(frame.rotation.subtract(this.rotation).scale(easing.ease(progress)));
        Vec3 scale = this.scale.add(frame.scale.subtract(this.scale).scale(easing.ease(progress)));
        return new Frame(position, rotation, scale);
    }
}

File: Common/src/main/java/foundry/veil/anim/Frame.kt
package foundry.veil.anim

import foundry.veil.math.Easings.Easing
import net.minecraft.util.Mth
import net.minecraft.world.phys.Vec3

open class Frame(open var pos: Vec3) {

    fun interpolate(frame: Frame, progress: Float, easing: Easing): Frame {
        val x = Mth.lerp(easing.ease(progress).toDouble(), pos.x, frame.pos.x)
        val y = Mth.lerp(easing.ease(progress).toDouble(), pos.y, frame.pos.y)
        val z = Mth.lerp(easing.ease(progress).toDouble(), pos.z, frame.pos.z)
        return Frame(Vec3(x, y, z))
    }

    fun x(): Double {
        return pos.x
    }

    fun y(): Double {
        return pos.y
    }

    fun z(): Double {
        return pos.z
    }
}